lib/src/main/java/uk/gov/di/ipv/cri/passport/library/validation/JarValidator.java

```diff
@@ -0,0 +1,187 @@
+            } else {
+                concatSignatureJwt = signedJWT;
+            }
+            boolean valid =
+                    concatSignatureJwt.verify(
+                            new ECDSAVerifier(
+                                    configurationService.getClientSigningPublicJwk(clientId)));
+
+            if (!valid) {
+                LOGGER.error("JWT signature validation failed");
+                throw new SharedAttributesValidationException(
+                        OAuth2Error.INVALID_REQUEST_OBJECT.setDescription(
+                                "JWT signature validation failed"));
+            }
+        } catch (JOSEException | ParseException e) {
+            LOGGER.error("Failed to parse JWT when attempting signature validation");
+            throw new SharedAttributesValidationException(
+                    OAuth2Error.INVALID_REQUEST_OBJECT.setDescription(
+                            "Failed to parse JWT when attempting signature validation"));
+        }
+    }
+
+    private JWTClaimsSet validateClaimSet(SignedJWT signedJWT, String clientId)
+            throws SharedAttributesValidationException {
+
+        String criAudience = configurationService.getAudienceForClients();
+        String clientIssuer = configurationService.getClientIssuer(clientId);
+
+        DefaultJWTClaimsVerifier<?> verifier =
+                new DefaultJWTClaimsVerifier<>(
+                        criAudience,
+                        new JWTClaimsSet.Builder()
+                                .issuer(clientIssuer)
+                                .claim("response_type", "code")
+                                .build(),
+                        new HashSet<>(
+                                Arrays.asList(
+                                        JWTClaimNames.EXPIRATION_TIME,
+                                        JWTClaimNames.NOT_BEFORE,
+                                        JWTClaimNames.ISSUED_AT,
+                                        JWTClaimNames.SUBJECT)));
+
+        try {
+            verifier.verify(signedJWT.getJWTClaimsSet(), null);
+
+            validateDateClaims(signedJWT.getJWTClaimsSet());
+
+            return signedJWT.getJWTClaimsSet();
+        } catch (BadJWTException | ParseException e) {
+            LOGGER.error("Claim set validation failed");
+            throw new SharedAttributesValidationException(
+                    OAuth2Error.INVALID_GRANT.setDescription(e.getMessage()));
+        }
+    }
+
+    private void validateDateClaims(JWTClaimsSet claimsSet)
+            throws SharedAttributesValidationException {
+        Date expirationTime = claimsSet.getExpirationTime();
+        String maxAllowedTtl = configurationService.getMaxClientAuthTokenTtl();
+
+        OffsetDateTime offsetDateTime =
+                OffsetDateTime.now().plusSeconds(Long.parseLong(maxAllowedTtl));
+        if (expirationTime.getTime() / 1000L > offsetDateTime.toEpochSecond()) {
```

Comment:
Merely a suggestion but I'd find the following easier to understand. 
```
        LocalDateTime maximumExpirationTime =
                LocalDateTime.now().plusSeconds(Long.parseLong(maxAllowedTtl));
        LocalDateTime expirationTime =
                LocalDateTime.ofInstant(claimsSet.getExpirationTime().toInstant(), ZoneOffset.UTC);

        if (expirationTime.isAfter(maximumExpirationTime)) {
```

Either way `claimset.getExpirationTime()` can return `null` if expiration time isn't set, if so you might want to check for null and handle it to avoid spurious NullPointerExceptions later on.

Reply:
Cheers I'll take a look at making this a bit clearer. It could probably do with a different method name aswell I think. Now its specifically only validating that the expiry date isn't too far in the future. It used to be doing all the date validation but realised that is covered in the DefaultJwtClaimsVerifier instead